### PR TITLE
Type annotations for sqlalchemy.orm.events

### PR DESCRIPTION
**Pull request by ghost**
### Description
An attempt to annotate `lib/sqlalchemy/orm/events.py` with type hints (issue #6810).

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 9a897b9ae746753ed5388c2c83c66ecd20d88f0e of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 9a897b9ae746753ed5388c2c83c66ecd20d88f0e: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

**Comment by zzzeek** [comment]
added to gerrit so I can track

**Comment by g-kisenkov** [comment]
@zzzeek 

Regarding the CI error `TypeError: 'staticmethod' object is not callable`. Turns out prior to Python 3.10, `staticmethod`s are not directly callable from inside the class. Should I just `# type: ignore` it or I can pull the method out in a stand-alone function?

**Comment by zzzeek** [comment]
> @zzzeek
> 
> Regarding the CI error `TypeError: 'staticmethod' object is not callable`. Turns out prior to Python 3.10, `staticmethod`s are not directly callable from inside the class and `mypy` doesn't catch it. Should I just remove the `@staticmethod` I put on `_lifecycle_event` and `# type: ignore` it, so that `mypy` wouldn't scream at `_lifecycle_event` not receiving class/instance argument or I can pull the method out in a stand-alone function?

yes, definitely remove the `@staticmethod`.  with these changes we should always opt to never change any compile-time or runtime behaviors, other than "if TYPE_CHECKING" / "cast()" / "assert x is not None"

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 9ae5cb80f46681f362b083ce6e446da60f17ecf3 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 9ae5cb80f46681f362b083ce6e446da60f17ecf3 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

File: lib/sqlalchemy/orm/events.py [hunk 1]
```diff
@@ -1730,7 +1864,12 @@ def do_something(session, previous_transaction):
 
         """
 
-    def before_flush(self, session, flush_context, instances):
+    def before_flush(
+        self,
+        session: Session,
+        flush_context: UOWTransaction,
+        instances: Optional[Sequence[Any]],
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/events.py:1871]: **mike bayer** (zzzeek) wrote:

instances is probably `Optional[Sequence[_O]]`

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

File: lib/sqlalchemy/orm/events.py [hunk 2]
```diff
@@ -2354,14 +2508,21 @@ def wrap(target, *arg, **kw):
...
                     mgr[target.key].dispatch._active_history = True
 
-    def append(self, target, value, initiator, *, key=NO_KEY):
+    def append(
+        self,
+        target: _O,
+        value: _T,
+        initiator: Event,
+        *,
+        key: EventConstants = NO_KEY,
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/events.py:2524]: **mike bayer** (zzzeek) wrote:

this is correct.

There's a change that's needed in sqlalchemy.orm.attributes to support this correctly which is at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4329 but that can remain independent of this.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

File: lib/sqlalchemy/orm/events.py [hunk 3]
```diff
@@ -2883,7 +3079,9 @@ def my_event(query):
 
         """
 
-    def before_compile_update(self, query, update_context):
+    def before_compile_update(
+        self, query: Query[Any], update_context: _O
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/events.py:3083]: **mike bayer** (zzzeek) wrote:

update_context here would be sqlalchemy.orm.query.BulkUpdate

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

File: lib/sqlalchemy/orm/events.py [hunk 4]
```diff
@@ -2933,7 +3131,9 @@ def no_deleted(query, update_context):
 
         """
 
-    def before_compile_delete(self, query, delete_context):
+    def before_compile_delete(
+        self, query: Query[Any], delete_context: _O
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/events.py:3135]: **mike bayer** (zzzeek) wrote:

delete_context here would be sqlalchemy.orm.query.BulkDelete

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

just about ready.  small fixes

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision a3fd2c0c3790164c433305ccc7ac6b73e813e037 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a3fd2c0c3790164c433305ccc7ac6b73e813e037 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

this is done, thanks very much!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326 has been **merged**. Congratulations! :)

**Comment by g-kisenkov** [comment]
> **mike bayer** (zzzeek) wrote:
> 
> this is done, thanks very much!
> 
> View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4326

Thank you for the review!

**Pull request closed without merging**